src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionOptions.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.IO.Pipelines;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
+{
+    /// <summary>
+    /// Options for socket based connections.
+    /// </summary>
+    public class SocketConnectionOptions
+    {
+        /// <summary>
+        /// The <see cref="PipeOptions"/> for socket connections used for input.
+        /// </summary>
+        public PipeOptions InputOptions { get; init; } = new PipeOptions();
+
+        /// <summary>
+        /// The <see cref="PipeOptions"/> for socket connections used for output.
+        /// </summary>
+        public PipeOptions OutputOptions { get; init; } = new PipeOptions();
+
+        /// <summary>
+        /// Delay socket read/write operations until the transport pipe is read from or written to.
+        /// </summary>
+        /// <remarks>
+        /// Defaults to false.
+        /// </remarks>
+        public bool DeferFirstOperation { get; set; }
+
+        /// <summary>
+        /// Wait until there is data available to allocate a buffer. Setting this to false can increase throughput at the cost of increased memory usage.
+        /// </summary>
+        /// <remarks>
+        /// Defaults to true.
+        /// </remarks>
+        public bool WaitForDataBeforeAllocatingBuffer { get; set; } = true;
+
+        internal MemoryPool<byte> MemoryPool { get; init; } = default!;
+
+        internal SocketSenderPool SenderPool { get; init; } = default!;
+
+        internal PipeScheduler Scheduler { get; init; } = default!;
+
+        internal ISocketsTrace Trace { get; init; } = default!;
```

Comment:
It seems wrong to mark these as non-nullable. Any non-internal user of this API will leave these null, right?

Reply:
I don't like having internal properties at all. External callers `ISocketConnectionContextFactory` won't be able to use these. The reason for making this a service instead of just a public static method is so these pools could be maintained internally by the SocketConnectionContextFactory.

Reply:
Yeah these internal properties feel weird, they are only here for us (and they should never be null when used), totally agree that this feels a bit yuck.  Any ideas on an alternative to shuttle this info?

Reply:
Our built-in stuff should just continue new up SocketConnection directly so we can pass in these internal types. The SocketConnectionContextFactory can create an `ISocketsTrace` from an `ILoggerFactory` that's constructor injected like SocketTransportFactory does.

Reply:
For the pools, the SocketConnectionContextFactory can just keep these in private fields. The reason it's a service and not a static method is not so it can be injested by the socket transport, but so it can keep these pools around between calls.